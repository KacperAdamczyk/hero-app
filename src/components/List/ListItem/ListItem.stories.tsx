/** @jsx jsx */
import { ReactNode } from 'react';
import { jsx, css } from '@emotion/core';
import { action } from '@storybook/addon-actions';

import { ListItem } from './ListItem';
import { Layout } from '../List';
import { Data } from '../List.stories';

export default {
  title: 'List / List item',
  component: ListItem,
  decorators: [
    (storyFn: () => ReactNode) => (
      <div
        css={css`
          background-color: lightblue;
          height: 100px;
          display: flex;
          align-items: center;
        `}
      >
        {storyFn()}
      </div>
    ),
  ],
};

const data: Data = {
  id: '1',
  name: 'Foo',
  type: 'Type 1',
  age: 42,
};

const layout: Layout<Data>[] = [
  {
    header: 'Header 1',
    space: 30,
    content: ({ name }) => <div>{name}</div>,
  },
  {
    header: 'Header 2',
    space: 20,
    content: ({ type }) => <div>{type}</div>,
  },
  {
    header: 'Header 3',
    space: 50,
    content: ({ age }) => <div>{age}</div>,
  },
];

export const Default = () => (
  <ListItem data={data} layout={layout} onClick={action('onClick')} />
);
