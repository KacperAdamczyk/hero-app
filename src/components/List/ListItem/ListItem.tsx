import React, { FC, PropsWithChildren, useCallback } from 'react';
import styled from '@emotion/styled';

import { Layout } from '../List';
import { media } from 'styling';

interface Props<T> {
  data: T;
  layout: Layout<T>[];
  onClick?: (data: T) => void;
}

const Container = styled.div`
  display: flex;
  box-sizing: border-box;
  min-height: 75px;
  padding: 20px 0;
  width: 100%;
  background-color: white;
  border-radius: 5px;
  align-items: center;
  cursor: pointer;

  ${media.small} {
    flex-direction: column;
    align-items: flex-start;
  }
`;

interface CellProps {
  space: number;
}

const Cell = styled.div<CellProps>`
  flex: ${({ space }) => space};
  max-width: ${({ space }) => space}%;
  box-sizing: border-box;
  padding: 0 5px;

  &:first-of-type {
    padding-left: 20px;
  }

  &:last-of-type {
    padding-right: 20px;
  }

  ${media.small} {
    padding-bottom: 20px;

    &:first-of-type {
      padding-right: 20px;
    }

    &:last-of-type {
      padding-left: 20px;
      padding-bottom: 0;
    }
  }
`;

export const ListItem = <T,>({
  data,
  layout,
  onClick,
}: PropsWithChildren<Props<T>>): ReturnType<FC> => {
  const onContainerClick = useCallback(() => onClick?.(data), [onClick, data]);

  return (
    <Container onClick={onContainerClick}>
      {layout.map(({ space, content }, index) => (
        <Cell key={index} space={space}>
          {content(data)}
        </Cell>
      ))}
    </Container>
  );
};
