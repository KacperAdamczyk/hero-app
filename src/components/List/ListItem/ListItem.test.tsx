import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { ListItem } from './ListItem';

describe('ListItem', () => {
  test('should fire an event on click', () => {
    const onClick = jest.fn();
    const data = 'foo';
    const { container, getByTestId } = render(
      <ListItem onClick={onClick} data={data} layout={[]} />,
    );

    fireEvent.click(getByTestId('container'));

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith('foo');
    expect(container.firstChild).toMatchSnapshot();
  });
});
