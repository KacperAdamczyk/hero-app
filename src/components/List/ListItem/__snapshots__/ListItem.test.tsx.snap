// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem should fire an event on click 1`] = `
<div
  class="css-1ttdpxg"
  data-testid="container"
/>
`;
